web/resource/resource.go

```diff
@@ -0,0 +1,178 @@
+// Package resource implements a RESTful resource handler
+package resource
+
+import (
+	"net/http"
+	"strings"
+
+	"darvaza.org/core"
+	"darvaza.org/x/web"
+	"darvaza.org/x/web/qlist"
+)
+
+var (
+	_ http.Handler = (*Resource[any])(nil)
+	_ web.Handler  = (*Resource[any])(nil)
+)
+
+// Resource is an http.Handler built around a given object
+// and a data type.
+type Resource[T any] struct {
+	check    CheckerFunc[T]
+	methods  []string
+	identity string
+
+	h  map[string]HandlerFunc[T]
+	r  map[string]HandlerFunc[T]
+	ql []qlist.QualityValue
+}
+
+// ServeHTTP handles the request initially using the TryServeHTTP method, and
+// then calling [web.HandleError] if there is any problem.
+func (r *Resource[T]) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if err := r.TryServeHTTP(rw, req); err != nil {
+		web.HandleError(rw, req, err)
+	}
+}
+
+// TryServeHTTP attempts to handle the request, and returns an error in the case of
+// problems instead of handling it locally.
+// This method converts panics into errors in case they occur.
+func (r *Resource[T]) TryServeHTTP(rw http.ResponseWriter, req *http.Request) error {
+	var c core.Catcher
+
+	return c.Do(func() error {
+		return r.doServeHTTP(rw, req)
+	})
+}
+
+func (r *Resource[T]) doServeHTTP(rw http.ResponseWriter, req *http.Request) error {
+	req, data, err := r.checkRequest(req)
+	if err != nil {
+		return err
+	}
+
+	h := r.getHandler(req)
+	if h == nil {
+		h = r.serve405
```

Comment:
This might be only philosophical, but here I see a `r.serve400`, not a `r.server405`. The problem here is that the request is malformed (as in it does not have a valid method) not that the method is not allowed/not appropriate for the resource. In this case a http 400 is beter than a http 405.

Reply:
a malformed request should be caught by the checker. calls for unsupported methods are not malformed requests. Checker is also in charge of 404 after validating the route parameters

Reply:
`h==nil` means no method for me, am I completely mistaken?

Reply:
`h` is the registered method handler, from the `getHandler()` on the line before.
we reach this point after a successful Check() for 400 and 404 errors to check if the resource supports the requested HTTP method.
if the resource doesn't the, HTTP error is _405 method not allowed_ which in this case it's _served_ by a handler that returns such 405 error with the required headers.

@karasz

Reply:
`h == nil` means no *method handler*, but there is a *method* on the request.

Reply:
I'll refactor this to make it more obvious @karasz

Reply:
PR amended, please re-review